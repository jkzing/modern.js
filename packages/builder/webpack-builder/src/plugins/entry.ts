import type { EntryObject } from 'webpack';
import type { BuilderPlugin } from '../types';

export const PluginEntry = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-entry',

  setup(api) {
    api.modifyWebpackChain(async chain => {
      const { ensureArray } = await import('@modern-js/utils');

      const { entry } = api.context;
      const config = api.getBuilderConfig();

      const { preEntry, vendorEntry = {} } = config.source || {};
      const preEntries = preEntry ? ensureArray(preEntry) : [];

      const addEntries = (entryObject: EntryObject) => {
        Object.keys(entryObject).forEach(entryName => {
          preEntries.forEach(entry => {
            chain.entry(entryName).add(entry);
          });

          ensureArray(entryObject[entryName]).forEach(item => {
            chain.entry(entryName).add(item as string);
          });
        });
      };
      addEntries(entry);
      addEntries(vendorEntry);
    });
  },
});
