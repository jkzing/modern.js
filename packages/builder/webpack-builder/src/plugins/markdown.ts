import type { BuilderPlugin } from '../types';

export const PluginMarkdown = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-markdown',

  setup(api) {
    api.modifyWebpackChain((chain, { CHAIN_ID, getCompiledPath }) => {
      chain.module
        .rule(CHAIN_ID.RULE.MARKDOWN)
        .test(/\.md$/)
        .use(CHAIN_ID.USE.HTML)
        .loader(require.resolve('html-loader'))
        .end()
        .use(CHAIN_ID.USE.MARKDOWN)
        .loader(getCompiledPath('markdown-loader'));
    });
  },
});
