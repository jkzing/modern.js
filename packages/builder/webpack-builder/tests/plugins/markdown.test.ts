import { expect, describe, it } from 'vitest';
import { PluginMarkdown } from '../../src/plugins/markdown';
import { createStubBuilder } from '../utils/builder';

describe('plugins/markdown', () => {
  it('should add markdown rules properly', async () => {
    const builder = createStubBuilder({
      plugins: [PluginMarkdown()],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
