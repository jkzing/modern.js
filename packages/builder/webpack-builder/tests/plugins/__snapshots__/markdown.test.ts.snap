// Vitest Snapshot v1

exports[`plugins/markdown > should add markdown rules properly 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.md\\$/,
        "use": [
          {
            "loader": "<NODE_MODULES>/html-loader/dist/cjs.js",
          },
          {
            "loader": "<ROOT>/packages/builder/webpack-builder/compiled/markdown-loader",
          },
        ],
      },
    ],
  },
}
`;
